src/llvm-cpufeatures.cpp

```diff
@@ -62,15 +62,26 @@ static bool have_fma(Function &intr, Function &caller, const Triple &TT) JL_NOTS
     SmallVector<StringRef, 128> Features;
     FS.split(Features, ',');
     for (StringRef Feature : Features)
-    if (TT.isARM()) {
-      if (Feature == "+vfp4")
-        return typ == "f32" || typ == "f64";
-      else if (Feature == "+vfp4sp")
-        return typ == "f32";
-    } else if (TT.isX86()) {
-      if (Feature == "+fma" || Feature == "+fma4")
-        return typ == "f32" || typ == "f64";
-    }
+        if (TT.isARM()) {
+            if (Feature == "+vfp4")
+                return typ == "f32" || typ == "f64";
+            else if (Feature == "+vfp4sp")
+                return typ == "f32";
+        } else if (TT.isX86()) {
+            if (Feature == "+fma" || Feature == "+fma4")
+                return typ == "f32" || typ == "f64";
+        } else if (TT.isRISCV64()) {
+            // Don't return early if we find a known extension but it doesn't
+            // support the current type (e.g. f64 with Zfh or f16 with D)
+            if (Feature == "+zfh" && (typ == "f16" || typ == "f32"))
+                // Zfh implies F: https://github.com/riscv/riscv-isa-manual/blob/9799e5b43fa6780e58ec440774b0194debbca52b/src/zfh.adoc
+                return true;
+            else if (Feature == "+d" && (typ == "f32" || typ == "f64"))
+                // D implies F: https://github.com/riscv/riscv-isa-manual/blob/9799e5b43fa6780e58ec440774b0194debbca52b/src/d-st-ext.adoc
```

Comment:
why not just have `Feature == "+zfh" && typ == "f16"`? Seems like relying on the extensions that imply other extensions is more complicated than necessary.

Reply:
Fair, that could save some cycles depending on the order of iteration of the features, but probably not worth the effort.